### PR TITLE
Set environment variables for go test runs

### LINKED ISSUE #475: How to set env var before run Go test command
hi, Need use env var in go test , So  I want  to add Environment Variables just for go test command.
how to do it ?  thanks!
`Go: Run test at cursor`

### PR DESCRIPTION
**Pull request by ramya-rao-a**
Fixes #475

### CODE REVIEW

File: src/goTest.ts [hunk 1]
```diff
@@ -94,19 +94,19 @@ export function testCurrentPackage(timeout: string) {
...
 		return;
 	}
-	getTestFunctions(editor.document).then(testFunctions => {
+	return getTestFunctions(editor.document).then(testFunctions => {
 		return goTest({
 			timeout: timeout,
 			dir: path.dirname(editor.document.fileName),
 			functions: testFunctions.map(func => { return func.name; })
-		});
-	}).then(null, err => {
```
Comment (lukehoban) [inline src/goTest.ts]: Possibly minor - but this change will mean that any exceptions thrown during execution of `goTest` are not handled and reported to `console.error`.  A weirdness of `.then` on promises.  Possibly just leave the additional `.then` call.

File: src/goTest.ts [hunk 2]
```diff
@@ -116,19 +116,23 @@ export function testCurrentFile(timeout: string) {
  *
  * @param config the test execution configuration.
  */
-function goTest(config: TestConfig): Thenable<boolean> {
+function goTest(config: TestConfig, goConfig?: vscode.WorkspaceConfiguration): Thenable<boolean> {
```
Comment (lukehoban) [inline src/goTest.ts]: Any reason not to make this required and fix up callers?

File: src/goTest.ts [hunk 3]
```diff
@@ -116,19 +116,23 @@ export function testCurrentFile(timeout: string) {
...
 		outputChannel.show(2);
-		let buildFlags: string[] = vscode.workspace.getConfiguration('go')['buildFlags'];
-		let buildTags: string = vscode.workspace.getConfiguration('go')['buildTags'];
+		if (!goConfig) {
+			goConfig = vscode.workspace.getConfiguration('go');
+		}
+		let buildFlags: string[] = goConfig['buildFlags'];
+		let buildTags: string = goConfig['buildTags'];
 		let args = ['test', '-v', '-timeout', config.timeout, '-tags', buildTags, ...buildFlags];
+		let testEnvVars = Object.assign({}, goConfig['testEnvVars'], process.env);
```
Comment (lukehoban) [inline src/goTest.ts]: Doesn't the order need to be changed to ensure that the `testEnvVars` overrride the process environment settings?

File: src/goTest.ts (hunk 1 above)
Comment (ramya-rao-a) [inline src/goTest.ts]: Put the `console.error(err)` to back where it was and returned promise from there

File: src/goTest.ts (hunk 2 above)
Comment (ramya-rao-a) [inline src/goTest.ts]: None at all. Made the update to pass goConfig from goMain.ts

File: src/goTest.ts (hunk 3 above)
Comment (ramya-rao-a) [inline src/goTest.ts]: True, nice catch. Corrected

**Comment by rynop** [comment]
For those trying to figure out how to use this, you have to set a value in your vs code `settings.json`:

Ex:
`"go.testEnvFile": "${workspaceFolder}/.env",`

**Comment by henrahmagix** [comment]
Or
```json
"go.testEnvVars": {
  "APP_DEBUG": "1",
  "MY_ENV": "my-value"
},
```

**Pull request merged**